### ISSUE TITLE
Stats performance degradation and overflows

### ISSUE THREAD

**Issue by maichmueller** [sympy/sympy#20521]
Hello,

since the move to sympy 1.7 there are multiple issues with the stats package of sympy, which have not been there in 1.6.2 (last I succesfully used). 
For one, the sampling speed is now decisively reduced. These are timings I get for 1.7 vs 1.6.2 with python 3.8.5:

1.7:
```
tic = perf_counter()
list(sample_iter(Normal("N", 0, 1), numsamples=1000))
toc = perf_counter()
print(toc - tic)
20.0278042
```
1.6.2:
```
tic = perf_counter()
list(sample_iter(Normal("N", 0, 1), numsamples=1000))
toc = perf_counter()
print(toc - tic)
0.10106169999998826
```
I suppose this is due to relying on scipy now by default, however such performance degradation is too much. Could it be that the sample_iter makes a call to scipy anew for every single sample? This would explain in part why this happens.

My package (SCM) now also fails in its tests due to sudden overflows in sampling (to be found here https://github.com/maichmueller/scm/runs/1479975310?check_suite_focus=true):
```
scm/scm.py:207: in sample
    list(sample_iter(node_attr[self.noise_key], numsamples=n)), dtype=float
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/sympy-1.7-py3.7.egg/sympy/stats/rv.py:1216: in return_generator
    d = ps.sample(size=size, library=library)  # a dictionary that maps RVs to values
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/sympy-1.7-py3.7.egg/sympy/stats/crv.py:601: in sample
    return {self.value: self.distribution.sample(size, library=library)}
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/sympy-1.7-py3.7.egg/sympy/stats/crv.py:290: in sample
    samps = _get_sample_class_crv[library](self, size)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/sympy-1.7-py3.7.egg/sympy/stats/crv.py:151: in __new__
    return cls._sample_scipy(dist, size)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/sympy-1.7-py3.7.egg/sympy/stats/crv.py:165: in _sample_scipy
    return scipy_rv.rvs(size=size)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/scipy-1.5.4-py3.7-linux-x86_64.egg/scipy/stats/_distn_infrastructure.py:1004: in rvs
    vals = self._rvs(*args, size=size, random_state=random_state)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/scipy-1.5.4-py3.7-linux-x86_64.egg/scipy/stats/_distn_infrastructure.py:923: in _rvs
    Y = self._ppf(U, *args)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/scipy-1.5.4-py3.7-linux-x86_64.egg/scipy/stats/_distn_infrastructure.py:938: in _ppf
    return self._ppfvec(q, *args)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/numpy-1.19.4-py3.7-linux-x86_64.egg/numpy/lib/function_base.py:2108: in __call__
    return self._vectorize_call(func=func, args=vargs)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/numpy-1.19.4-py3.7-linux-x86_64.egg/numpy/lib/function_base.py:2192: in _vectorize_call
    outputs = ufunc(*inputs)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/scipy-1.5.4-py3.7-linux-x86_64.egg/scipy/stats/_distn_infrastructure.py:1694: in _ppf_single
    while self._ppf_to_solve(right, q, *args) < 0.:
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/scipy-1.5.4-py3.7-linux-x86_64.egg/scipy/stats/_distn_infrastructure.py:1679: in _ppf_to_solve
    return self.cdf(*(x, )+args)-q
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/scipy-1.5.4-py3.7-linux-x86_64.egg/scipy/stats/_distn_infrastructure.py:1854: in cdf
    place(output, cond, self._cdf(*goodargs))
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/scipy-1.5.4-py3.7-linux-x86_64.egg/scipy/stats/_distn_infrastructure.py:1729: in _cdf
    return self._cdfvec(x, *args)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/numpy-1.19.4-py3.7-linux-x86_64.egg/numpy/lib/function_base.py:2108: in __call__
    return self._vectorize_call(func=func, args=vargs)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/numpy-1.19.4-py3.7-linux-x86_64.egg/numpy/lib/function_base.py:2192: in _vectorize_call
    outputs = ufunc(*inputs)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/scipy-1.5.4-py3.7-linux-x86_64.egg/scipy/stats/_distn_infrastructure.py:1726: in _cdf_single
    return integrate.quad(self._pdf, _a, x, args=args)[0]
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/scipy-1.5.4-py3.7-linux-x86_64.egg/scipy/integrate/quadpack.py:352: in quad
    points)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/scipy-1.5.4-py3.7-linux-x86_64.egg/scipy/integrate/quadpack.py:465: in _quad
    return _quadpack._qagie(func,bound,infbounds,args,full_output,epsabs,epsrel,limit)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/sympy-1.7-py3.7.egg/sympy/stats/crv.py:162: in _pdf
    return handmade_pdf(x)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

Dummy_34449 = 1.0000000000000005e+155

    def _lambdifygenerated(Dummy_34449):
>       return ((1/4)*sqrt(2)*exp(-1/8*(Dummy_34449 - 5)**2)/sqrt(pi))
E       OverflowError: (34, 'Numerical result out of range')

<lambdifygenerated-17255>:2: OverflowError
```

This is also related to scipy and did not occur in 1.6.2 either.
Can somebody point me towards how to use the new stats package correctly?

**Comment by oscarbenjamin** [comment]
CC @Smit-create @czgdp1807 @Upabjojr 

It looks like the speed issue is already fixed on master:
```python
In [1]: from sympy.stats import *

In [2]: %time _ = list(sample_iter(Normal("N", 0, 1), numsamples=1000))
CPU times: user 901 ms, sys: 127 ms, total: 1.03 s
Wall time: 592 ms

In [3]: %time _ = list(sample_iter(Normal("N", 0, 1), numsamples=1000))
CPU times: user 123 ms, sys: 6.52 ms, total: 129 ms
Wall time: 124 ms
```
Not sure what the fix was but I guess it could be backported to the 1.7 branch so it could go into 1.7.1 as a bugfix.

Can you show how to reproduce the overflow error?

**Comment by naveensaigit** [comment]
> Can you show how to reproduce the overflow error?

I'm not sure about this, but I faced a similar issue once a few days back when I was testing the code locally. I tried to reproduce it, but couldn't do so. I think it might have something to do with random state if sampling uses it.

**Comment by maichmueller** [comment]
> Not sure what the fix was but I guess it could be backported to the 1.7 branch so it could go into 1.7.1 as a bugfix.

That would be well appreciated. The speed issue breaking the usefulness of my package.

> Can you show how to reproduce the overflow error?

This does it at random (increasing sample size probably increases its occurence)
```
from sympy.stats import Normal, sample_iter
list(sample_iter( Normal("N", 5, 2), numsamples=100))
[5.324059585058186, ...]
list(sample_iter( Normal("N", 5, 2), numsamples=100))
[8.453554584921594, ...]
list(sample_iter( Normal("N", 5, 2), numsamples=100))
[3.185947733128358, ...]
list(sample_iter( Normal("N", 5, 2), numsamples=100))
[4.038455094645867, ...]
list(sample_iter( Normal("N", 5, 2), numsamples=100))
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\sympy\stats\rv.py", line 1216, in return_generator
    d = ps.sample(size=size, library=library)  # a dictionary that maps RVs to values
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\sympy\stats\crv.py", line 601, in sample
    return {self.value: self.distribution.sample(size, library=library)}
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\sympy\stats\crv.py", line 290, in sample
    samps = _get_sample_class_crv[library](self, size)
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\sympy\stats\crv.py", line 151, in __new__
    return cls._sample_scipy(dist, size)
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\sympy\stats\crv.py", line 165, in _sample_scipy
    return scipy_rv.rvs(size=size)
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 1004, in rvs
    vals = self._rvs(*args, size=size, random_state=random_state)
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 923, in _rvs
    Y = self._ppf(U, *args)
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 938, in _ppf
    return self._ppfvec(q, *args)
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\numpy\lib\function_base.py", line 2108, in __call__
    return self._vectorize_call(func=func, args=vargs)
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\numpy\lib\function_base.py", line 2192, in _vectorize_call
    outputs = ufunc(*inputs)
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 1694, in _ppf_single
    while self._ppf_to_solve(right, q, *args) < 0.:
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 1679, in _ppf_to_solve
    return self.cdf(*(x, )+args)-q
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 1854, in cdf
    place(output, cond, self._cdf(*goodargs))
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 1729, in _cdf
    return self._cdfvec(x, *args)
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\numpy\lib\function_base.py", line 2108, in __call__
    return self._vectorize_call(func=func, args=vargs)
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\numpy\lib\function_base.py", line 2192, in _vectorize_call
    outputs = ufunc(*inputs)
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 1726, in _cdf_single
    return integrate.quad(self._pdf, _a, x, args=args)[0]
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\scipy\integrate\quadpack.py", line 351, in quad
    retval = _quad(func, a, b, args, full_output, epsabs, epsrel, limit,
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\scipy\integrate\quadpack.py", line 465, in _quad
    return _quadpack._qagie(func,bound,infbounds,args,full_output,epsabs,epsrel,limit)
  File "C:\Users\Michael\.conda\envs\py38\lib\site-packages\sympy\stats\crv.py", line 162, in _pdf
    return handmade_pdf(x)
  File "<lambdifygenerated-413>", line 2, in _lambdifygenerated
    return ((1/4)*sqrt(2)*exp(-1/8*(Dummy_837 - 5)**2)/sqrt(pi))
OverflowError: (34, 'Result too large')
```

**Comment by czgdp1807** [comment]
Well the fix was usage of dictionary instead of lamdify in the following class, https://github.com/sympy/sympy/blob/e036b1cfa1157b796b19ed0a1a0af97d40269366/sympy/stats/crv.py#L148

**Comment by czgdp1807** [comment]
https://github.com/sympy/sympy/pull/20417/ is the PR which did that.

**Comment by oscarbenjamin** [comment]
Does that fix both problems?

The traceback above indicates that the failure is in quadpack which I assume is for numerical integration. Why is it using numerical integration for the normal distribution?

**Comment by czgdp1807** [comment]
Earlier we lamdified the PDF and gave that to scipy. Now scipy treated it as a custom distribution with some PDF. Probably, it is using https://en.wikipedia.org/wiki/Inverse_transform_sampling#The_method and it is integrating the PDF to obtain CDF/inverse-CDF. Anyways, how `scipy` samples isn't `sympy`'s concern. The root cause of random failure earlier was not using the in-built distribution of `scipy` and providing a PDF which `scipy` treated as custom distribution.

 > Does that fix both problems?

I haven't applied the diff of https://github.com/sympy/sympy/pull/20417 to `1.7`.

**Comment by oscarbenjamin** [comment]
> how `scipy` samples isn't `sympy`'s concern

It's worth considering whether it would be better to compute the inverse-CDF symbolically in sympy rather than letting scipy do it numerically. Looking in the docs for `rv_continuous` I see:
```
Correct, but potentially slow defaults exist for the remaining methods but for speed
and/or accuracy you can over-ride:

_logpdf, _cdf, _logcdf, _ppf, _rvs, _isf, _sf, _logsf

The default method _rvs relies on the inverse of the cdf, _ppf, applied to a uniform
random variate. In order to generate random variates efficiently, either the default
_ppf needs to be overwritten (e.g. if the inverse cdf can expressed in an explicit
form) or a sampling method needs to be implemented in a custom _rvs method.
```
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.html

Of course there is a better approach in the case of the normal distribution but seeing how scipy apparently struggles with sampling that from a pdf similar considerations might apply to any other hand made distribution.

You can see that sympy can easily calculate the inverse CDF in this case:
```
In [16]: pdf = exp(-x**2)/sqrt(pi)

In [17]: cdf = integrate(exp(-x**2)/sqrt(pi), (x, -oo, x))

In [18]: cdf
Out[18]: 
erf(x)   1
────── + ─
  2      2

In [19]: (ppf,) = solveset(cdf - y, x)

In [20]: ppf
Out[20]: erfinv(2⋅y - 1)
```
Then we can make a much more efficient custom sampler with scipy:
```
In [21]: lambdify(y, ppf)
Out[21]: <function _lambdifygenerated(y)>

In [22]: f_ppf = lambdify(y, ppf)

In [23]: from scipy.stats import rv_continuous

In [24]: class gaussian_gen(rv_continuous):
    ...:     def _ppf(self, y):
    ...:         return f_ppf(y)

In [25]: gaussian = gaussian_gen(name='gaussian')

In [26]: %time _ = gaussian().rvs(1000000)
CPU times: user 53.6 ms, sys: 12 ms, total: 65.6 ms
Wall time: 64.7 ms
```
That's more than 100000x faster than defining the pdf instead:
```
In [81]: f_pdf = lambdify(x, pdf)

In [82]: class gaussian_gen(rv_continuous):
    ...:     def _pdf(self, x):
    ...:         return f_pdf(x)
    ...: 

In [83]: gaussian = gaussian_gen(name='gaussian')

In [84]: %time _ = gaussian().rvs(10)
CPU times: user 106 ms, sys: 3.92 ms, total: 110 ms
Wall time: 107 ms
```
Providing the CDF alone is about 4x faster than providing only the pdf:
```
In [96]: %time _ = gaussian().rvs(100)
CPU times: user 269 ms, sys: 31.1 ms, total: 300 ms
Wall time: 275 ms
```

**Comment by czgdp1807** [comment]
That is I think a long term thing to consider whether `sympy.stats` should have it's own sampling algorithm in addition to support for external libraries or how the current method of linking with external libraries can be improved further. In fact, it's a project in itself, because it needs extensive discussions and some profiling before we move on with the implementation. We can surely add it to our future plans(may be part of a GSoC project) if it receives enough support from the community. Thanks for the profiling stats in your comment. They are worth considering. @Upabjojr What do you say?
But yeah the random failures are very much fixed by the usage of RV maps for `scipy`. So, they can be added to 1.7 as well.

**Comment by oscarbenjamin** [comment]
> I haven't applied the diff of #20417 to `1.7` yet, so can't say for sure.

I'm preparing for a 1.7.1 release. Have you been working on backporting this fix?

**Comment by czgdp1807** [comment]
I will open a PR shortly.

**Comment by Maelstrom6** [comment]
I can do PRs for drv and frv that matches #20417 today or tomorrow. Though, the main problem was with continuous RVs I think.

Mapping to SciPy's distributions will be far faster than trying to solve our own inverse. Many distributions do not have a closed form inverse cdf.

**Comment by czgdp1807** [comment]
> I can do PRs for drv and frv that matches #20417 today or tomorrow. Though, the main problem was with continuous RVs I think.

We should do it after #20494. The regression issue is fixed. We will do it for others after #20494.

**Comment by czgdp1807** [comment]
> Mapping to SciPy's distributions will be far faster than trying to solve our own inverse. Many distributions do not have a closed form inverse cdf.

There is no harm in adding distribution specific optimizations wherever necessary. No approach is perfect, which ever suits a distribution we should use it. @oscarbenjamin 's suggestion are good. We should do more profiling and take average results and compare it with the code in `master`. If they are good we can go for optimizing Normal distribution. Same goes for other distributions as well.

**Comment by Maelstrom6** [comment]
> oscarbenjamin 's suggestion are good. We should do more profiling and take average results and compare it with the code in master. If they are good we can go for optimizing Normal distribution. Same goes for other distributions as well.

Oh yes, my apologies, it will be useful for distributions that are not in the dictionary. I also raised [an issue on SciPy](https://github.com/scipy/scipy/issues/13051) and there is an algorithm called the ratio-of-uniforms which we could use instead of finding an inverse. It will be about as fast as the inverse method but we only require the maximum value of the pdf instead of having to find its inverse. The maximum value also doesn't need to be exact.

> > I can do PRs for drv and frv that matches #20417 today or tomorrow. Though, the main problem was with continuous RVs I think.
> 
> We should do it after #20494. The regression issue is fixed by adding scipy RV map to `1.7`. We will do it for others after #20494.

I see now that drv already has a map and frv doesn't need one. So no problems there.

**Comment by Maelstrom6** [comment]
@maichmueller If you are unable to wait a few days and need a fix immediately, I suggest having 
```python
next(sample(Normal("N", 0, 1), numsamples=1000, library="numpy"))
```
See a similar issue #20376.

**Comment by oscarbenjamin** [comment]
> it will be useful for distributions that are not in the dictionary.

I think it's worth bearing in mind that this is the most important use case for sympy in sampling. If the intention is to generate normally distributed random samples then users would be better served using scipy directly. What sympy can offer is the symbolic manipulation that could turn a custom analytically defined distribution into something that could be efficiently sampled from scipy.

**Comment by maichmueller** [comment]
> 
> 
> @maichmueller If you are unable to wait a few days and need a fix immediately, I suggest having
> 
> ```python
> next(sample(Normal("N", 0, 1), numsamples=1000, library="numpy"))
> ```
> 
> See a similar issue #20376.

I can wait. I brought this issue up, in order to guarantee my own pkg's usefulness. Good to see that the support here is very active!

**Comment by czgdp1807** [comment]
Random failures fixed by https://github.com/sympy/sympy/pull/20527
Efficiency will be improved in future versions because it needs some extensive discussions on advantages and disadvantages of various approaches for specific distributions. Closing this one and opening a new issue for further discussions.
